File: Design_Algo/Design/Algo_py/closestBSTValue.py
def findClosesValueInBst(tree , target):
    return findClosestValueInBstHelper(tree , target , float("inf"))

def findClosestValueInBstHelper(tree , target , closest):
    if tree is None:
        return closest
    if abs(target -closest) > abs(target - tree.value):
        closest = tree.value
    if target <tree.value:
        return findClosestValueInBstHelper(tree.left , target, closest)
    elif target > tree.value:
        return findClosestValueInBstHelper(tree.right, target , closest)
    else:
        return closest


def findClosesValueInBstIterative(tree , target):
    current = tree
    while current is not None:
        if abs(target -current.value) < abs(target - closest):
            closest = current.value
        if current.value - target < 0 :
            current = current.left
        elif current.value -target >0 :
            current = current.right
        else:
            return closest
            
    return closest

File: Design_Algo/Design/Algo/closestBSTValue.java
import java.util.*;
// O(log(N)) worst cases 0(N)   if recursively 0(d) ==> where d is depth 
class Program {
  public static int findClosestValueInBst(BST tree, int target) {
    int closest = Integer.MAX_VALUE;
    int current = tree;
    if(tree ==null){
        return -1;
    }
    while(current !=null){

        int diff = Math.abs(current.value -target);
        int oldClosestDiff = Math.abs(closest - target);
        if(diff < oldClosestDiff){
            closest = diff;
        }
        if(current.value - target < 0){
            current = current.left;
        }
        if(current.value - target > 0){
            current = current.right;
        }
        if(current.value - target ==0){
            return current.value;
        }
    }
    return closest;
  }

  static class BST {
    public int value;
    public BST left;
    public BST right;

    public BST(int value) {
      this.value = value;
    }
  }
}
